internal/pkg/agent/application/fleet_server_bootstrap.go

```diff
@@ -100,6 +108,60 @@ func FleetServerComponentModifier(serverCfg *configuration.FleetServerConfig) co
 	}
 }
 
+// injectESHosts will take the hosts list in src and add them to the hosts list in dst if they are not there.
+// Items in the hosts list will not have the http(s) schema prefix.
+// If src does not have an api_key attribute the function is a nop.
+// api_key attribute is missing when src is not retrieved from fleet.
+func injectESHosts(dst, src map[string]interface{}) error {
+	if _, ok := src["api_key"]; !ok {
+		return nil
+	}
+
+	// get src hosts as a string slice
+	srcAttr, ok := src["hosts"]
+	if !ok {
+		return fmt.Errorf("source does not contain hosts attribute")
+	}
+	srcList, ok := srcAttr.([]interface{})
+	if !ok {
+		return fmt.Errorf("source hosts attribute is not []interface{}: type=%T", srcAttr)
+	}
+
+	//get dst hosts as a string slice
+	dstAttr, ok := dst["hosts"]
+	if !ok {
+		return fmt.Errorf("destination does not contain hosts attribute")
+	}
+	dstList, ok := dstAttr.([]interface{})
+	if !ok {
+		return fmt.Errorf("destination hosts attribute is not []inteface{}: type=%T", dstAttr)
+	}
+
+	m := make(map[string]struct{}, len(srcList)+len(dstList))
+	for i, o := range srcList {
+		if host, ok := o.(string); ok {
+			host = schemaReg.ReplaceAllString(host, "")
```

Comment:
Why do you need a regex to do this? Could you use https://pkg.go.dev/net/url#Parse and then just override the `Scheme` member?

Reply:
I did a quick test of this; If you parsed an input that's `https://HOST`, and replaced `u.Schema` with an empty string it results in `//HOST`

Reply:
I had to remind myself about exactly what the problem was. We are in the right part of the code.

https://github.com/elastic/elastic-agent/issues/2784#issuecomment-1578741803

> Yes I forgot. Elasticsearch output configuration for the Fleet Server only every uses what it was bootstrapped with. It does not use the configuration that comes from the policy. It has been that way since it was implemented.

So this is the root of our problem and we need to decide how to fix this in a non-breaking way. The way it is designed now, the Fleet server output configuration is always overridden by the agent to be the output it was bootstrapped with:

https://github.com/elastic/elastic-agent/blob/c9bb1643c6817879545c9200e1d3d00c02cd3287/internal/pkg/agent/application/fleet_server_bootstrap.go#L67-L77

The simplest possible mitigation for this would be to remove that block completely, and just take the Fleet server outputs directly from the policy.

However I think this could be a breaking change for users who have created outputs that work for everything except Fleet server. That is, there could be somebody out there who is relying on the bootstrap output never changing and Fleet server ignoring the updated output in Fleet. We don't want to break their setups with an unfortunate surprise. For example, the output in Fleet could have a proxy configuration assuming its the output for Beats/Agent while fleet-server is deployed behind the proxy.

In particular if we look at the definition of a Fleet server output (which almost but not quite matches the [configuration in fleet-server](https://github.com/elastic/fleet-server/blob/fbb07ee59549179a3772149bf350ee2722705bb6/internal/pkg/config/output.go#L36) itself):

https://github.com/elastic/elastic-agent/blob/c9bb1643c6817879545c9200e1d3d00c02cd3287/internal/pkg/agent/configuration/fleet_server.go#L36-L48

What we are doing is extending the `hosts` list assuming that the rest of the bootstrap output configuration entries are still right (for example the TLS section). This isn't necessarily true, and at least if we keep doing it that way we've only allowed managing the `hosts` list from Fleet. Even though you can change more parameters than that in the UI they all get silently ignored. I don't think this is completely correct.

We have a similar situation to this in the agent when we introduced managing proxy parameters from Fleet **after* we had let users put them on the command line. The way we dealt with it was by letting the Fleet proxy replace the command line one only if we confirmed it still allowed us to connect to Fleet, otherwise [we fall back](https://github.com/elastic/elastic-agent/blob/c9bb1643c6817879545c9200e1d3d00c02cd3287/internal/pkg/agent/application/actions/handlers/handler_action_policy_change.go#L168-L197) to what was set before.

I think we'd want to do something similar here, where fleet-server uses the outputs set in the Fleet UI and ignores the bootstrap output completely only if confirms the outputs from Fleet UI work.

The trouble is to do this you have to send both outputs to Fleet server, and mark one as being the bootstrap output (or move the original bootstrap output under a `bootstrap` key in the single output configuration or something) so it can do the swap correctly.

I think that makes this more complicated, but I worry if we don't do it this way the implementation will be both incomplete (we ignore everything but `hosts`) and possibly break some users that implicitly depend on the outputs in Fleet UI being ignored by Fleet server.

Reply:
One thing to note is the output block modification:

https://github.com/elastic/elastic-agent/blob/c9bb1643c6817879545c9200e1d3d00c02cd3287/internal/pkg/agent/application/fleet_server_bootstrap.go#L67-L77
applies to fleet-server outputs only (and not any other component's elasticsearch output). It's required for bootstrapping where the initial "policy" from fleet-server that is used lists default values (which is why my modification method checks to see if an `api_key` is present as that indicates it's from fleet).

For example, if I add logging to the component modifier we can see the initial value of `unit.Config.Source.AsMap()` is:

```json
{
  "hosts":["localhost:9200"],
  "type":"elasticsearch"
}
```

And when it's retrieved from fleet we see:

```json
{
  "api_key":"REDACTED",
  "hosts":["https://HOST1:PORT","https://HOST2:PORT"],
  "preset":"balanced",
  "type":"elasticsearch"
}
```

If tls or proxy settings are available we see them in the object as well.

I don't think we want to send multiple output blocks to fleet-server as then fleet policies will need to be aware of things like mTLS usage and proxies.

I think we can adjust the output modification to assemble the output block using the retrieved policy and local settings.
In order to assemble the output on elastic-agent we can check if the policy from fleet contains tls/proxy settings and use the only if the keys are present; otherwise fallback to using the local settings.
For example if the local settings are:
```json
{
  "hosts":["HOST:9200"],
  "type":"elasticsearch",
  "ssl":{"certificate_authorities": "/path/to/ca.crt"},
  "proxy_url":"http://PROXY:PORT"
}
```
And it retrieves
```json
{
  "hosts":["OTHER:9200"],
  "type":"elasticsearch",
  "ssl":{"verification_mode": "none"}
}
```

It will create and test:
```json
{
  "hosts":["OTHER:9200"],
  "type":"elasticsearch",
  "ssl":{"verification_mode": "none"},
  "proxy_url":"http://PROXY:PORT"
}
```
If it's able to connect to elasticsearch the output block is passed to fleet-server; otherwise the block from local settings are used.
We would need to be able to persist the new block if it succeeds in order to properly handle restarts

Reply:
one question/thing to consider here. In the case the ES hosts in the policy aren't reachable and the agent keeps the cli configuration, what will be the expected behaviour? A unhealthy agent with a healthy fleet-server component?
Or as it's the bootstrap, it just fails all together?

Reply:
Another way to model this (and perhaps the way it was intended) is to treat it like a bootstrap broker in Kafka, where you manually provide a subset of the total available broker hosts which are used to discover the others.

In this situation, we wouldn't change anything for the agent. It would continue to pass Fleet server the bootstrap output. What would change is in Fleet server, where it internally looks up the other outputs configured for itself and merges them with the bootstrap output as long as doing that maintains connectivity to ES. 

With the changes in this PR, what happens is:

1. agent bootstraps fleet server
2. fleet server gets the agent policy from ES
3. fleet server sends that policy to agent
4. agent sees the policy has more outputs for fleet-server
5. agent merges them with the bootstrap output for fleet server
6. agent passes the merged output to fleet server.

We actually only need to get to step 2 for Fleet server to discover it has more possible outputs. I think keeping 3-6 here is unnecessary complexity and is probably putting the change in the wrong place (agent). If we push this change into Fleet server, it always has the policy outputs persisted in ES, and agent is always passing it the bootstrap output, so it always has all the information it needs to merge or switch between them as needed.

The only thing this doesn't do is make it possible for a user to change the bootstrap output from Fleet, but nobody is asking us to do that yet, and that isn't the change the issue here is requesting.